### PR TITLE
feat: integrate user segment with analytics explore

### PR DESCRIPTION
**Pull request by jingnanl**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

1. Integrate user segment with analytics explore
2. Fix segment UI bugs

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/24/2024, 03:54:40 (UTC+0)_

**Comment by zxkane** [comment]
# SonarQube Code Analytics
## Quality Gate passed

passed

## Additional information
*The following metrics might not affect the Quality Gate status but improving them will improve your project code quality.*

## Issues
bug   [0 Bugs](http://sonarqube:9000/project/issues?id=pr-1475&resolved=false&sinceLeakPeriod=true&types=BUG)

vulnerability   [0 Vulnerabilities](http://sonarqube:9000/project/issues?id=pr-1475&resolved=false&sinceLeakPeriod=true&types=VULNERABILITY)

code_smell   [0 Code Smells](http://sonarqube:9000/project/issues?id=pr-1475&resolved=false&sinceLeakPeriod=true&types=CODE_SMELL)

## Coverage and Duplications
*No data* **Coverage**

*No data* **Duplication**

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.80|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/24/2024, 04:25:39 (UTC+0)_

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :exclamation: Error
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:exclamation: Error|3|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:exclamation: Error|1|> 0|
|Code smells|:white_check_mark: OK|10|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:exclamation: Error|1|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/24/2024, 08:12:14 (UTC+0)_

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/25/2024, 06:49:32 (UTC+0)_

File: frontend/src/pages/analytics/segments/components/SegmentEditor.tsx [hunk 1]
```diff
@@ -407,7 +412,9 @@ const SegmentEditor: React.FC<SegmentEditorProps> = (
                   updateSegmentObject('expireDate', e.detail.value);
                   updateSegmentObject('refreshSchedule', {
                     ...segmentObject.refreshSchedule,
-                    expireAfter: new Date(e.detail.value).getTime(),
+                    expireAfter: e.detail.value
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/segments/components/SegmentEditor.tsx:415]: use ternary from `utils.ts`

File: frontend/src/pages/analytics/segments/UserSegmentDetails.tsx [hunk 2]
```diff
@@ -614,7 +616,7 @@ const getAutoRefreshExpiration = (
   }
 
   const expiration = segment?.refreshSchedule.expireAfter;
-  return getDateTimeWithTimezoneString(expiration, timezone);
+  return expiration ? getDateTimeWithTimezoneString(expiration, timezone) : '-';
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/segments/UserSegmentDetails.tsx:619]: use ternary from `utils.ts`

File: frontend/src/pages/analytics/analytics-utils.ts [hunk 3]
```diff
@@ -1371,3 +1384,34 @@ export const convertCronExpByTimeRange = (
...
+      categoryType: 'attribute',
+      itemList: [
+        {
+          category: ConditionCategoryFrontend.USER_OUTER,
+          description: categoryName,
+          label: itemName,
+          name: 'segment_id',
+          valueType: ExtendedMetadataValueType.USER_SEGMENT,
+          values: segmentGroupList.map((segment) => ({
+            value: segment.value ?? '',
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/analytics-utils.ts:1410]: use **defaultStr** from `utils.ts`

File: frontend/src/pages/analytics/segments/components/SegmentEditor.tsx (hunk 1 above)
Comment (jingnanl) [reply frontend/src/pages/analytics/segments/components/SegmentEditor.tsx]: done

File: frontend/src/pages/analytics/analytics-utils.ts (hunk 3 above)
Comment (jingnanl) [reply frontend/src/pages/analytics/analytics-utils.ts]: done

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.80|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/25/2024, 11:21:42 (UTC+0)_

File: src/control-plane/backend/lambda/api/router/segment.ts [hunk 4]
```diff
@@ -33,7 +33,7 @@ const segmentServ = new SegmentServ();
 router_segment.post(
   '',
   validate([
-    body().custom(isValidEmpty).custom(isXSSRequest),
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/router/segment.ts:36]: why removing this check?

**Review by chenhaiyun — APPROVED** [review]

File: src/control-plane/backend/lambda/api/router/segment.ts (hunk 4 above)
Comment (jingnanl) [reply src/control-plane/backend/lambda/api/router/segment.ts:36]: I found the segment request failed sometimes due to this check even the segment is pretty simple. So I remove it as it's unlikely to have XSS attack in create segment api request

Comment (jingnanl) [reply src/control-plane/backend/lambda/api/router/segment.ts]: I did more tests, but couldn't reproduce the above issue. Add the XSS check back.

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.80|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/26/2024, 01:31:27 (UTC+0)_

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @jingnanl on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|5.10|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1475)
###### _updated: 6/26/2024, 01:42:37 (UTC+0)_

**Review by zxkane — APPROVED** [review]

**Pull request merged**